Split TensorIteratorConfig out of TensorIterator

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39803 Split TensorIteratorConfig out of TensorIterator**

The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)
In fact, most of the stuff in TensorIteratorConfig isn't used by
TensorIterator later, so it can be dropped entirely after construction.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIterator iter = TensorIteratorConfig()
  .config1()
  .config2()
  .config3()
  .build();
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22018845](https://our.internmc.facebook.com/intern/diff/D22018845)

ezyang committed:
[WIP] Split TensorIteratorConfig out of TensorIterator

The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)

This diff doesn't compile, but I'm putting it up to get some early
feedback on the new API.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIteratorConfig config;
config.config1();
config.config2();
config.config3();
TensorIterator iter(config);
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Split TensorIteratorConfig out of TensorIterator"


The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)

This diff doesn't compile, but I'm putting it up to get some early
feedback on the new API.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIteratorConfig config;
config.config1();
config.config2();
config.config3();
TensorIterator iter(config);
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -744,15 +738,21 @@ TensorIterator TensorIterator::reduce_op(Tensor& out1, Tensor& out2, const Tenso
   iter.add_input(a);
-  iter.config_resize_outputs_ = false;
-  iter.config_is_reduction_ = true;
+  iter.dont_resize_outputs();
```
Aside: I didn't update "dont_resize_outputs" but for consistency it's probably better as "resize_outputs(bool)," or you can make the other config flags more consistent with it: "disable_resizing_outputs(), enable_promoting_inputs()..."

ezyang replied:
As preexisting conditions, I'm going to fix these in separate PRs.

mruberry commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -744,15 +738,21 @@ TensorIterator TensorIterator::reduce_op(Tensor& out1, Tensor& out2, const Tenso
-  iter.config_resize_outputs_ = false;
-  iter.config_is_reduction_ = true;
+  iter.dont_resize_outputs();
+  iter.is_reduction(true);
```
iter.make_reduction() and iter.make_elementwise()?

mruberry commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -744,15 +738,21 @@ TensorIterator TensorIterator::reduce_op(Tensor& out1, Tensor& out2, const Tenso
+  return TensorIterator(iter);
+}
+
+void TensorIterator::populate_operands() {
```
If you want to do something like this then you should require that the config be moved in the TI constructor, right? 
However, it'd be nice if you didn't have to copy into the TI at all. Would a pattern like config().build_iterator() -> TensorIterator be more natural?

ezyang replied:
It's actually OK, moving doesn't help at all because TensorIteratorConfig is effectively a value type (because SmallVector inlines its storage into the struct, moving doesn't save you any memory traffic, you still have to move each of the individual entries; everything is just primitive types).

> However, it'd be nice if you didn't have to copy into the TI at all. Would a pattern like config().build_iterator() -> TensorIterator be more natural?

Maybe. It depends whether or not the configuration options are used outside of build. I will investigate this; if we can avoid depending on the configuration options beyond build that actually is a pretty big improvement.

mruberry replied:
Sorry, I was being unclear. My suggestion to std::move the config into the TI if you follow this pattern was for safety.

ngimel replied:
Constructing/destructing `operands_` is a noticeable part of TensorIterator ops, so we should be careful not to do it twice, in config and in TI itself.

ezyang replied:
> My suggestion to std::move the config into the TI if you follow this pattern was for safety.

Yeah, I guess what I'm saying is that this doesn't actually help (because the move constructor on SmallVector is nearly just as expensive as the copy constructor).

> Constructing/destructing operands_ is a noticeable part of TensorIterator ops, so we should be careful not to do it twice, in config and in TI itself.

Duly noted

mruberry commented:
Made a couple comments. I like the idea a lot, but if building the Iterator is an atomic operation then maybe a different pattern will help you avoid copying the config into the iterator.

ezyang commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -680,57 +680,51 @@ TensorIterator TensorIterator::binary_op(Tensor& out, const Tensor& a,
 TensorIterator TensorIterator::unary_op(Tensor& out, const Tensor& a,
-    bool check_mem_overlap) {
-  auto iter = TensorIterator();
+    bool check_mem_overlapConfig) {
```
no this is a typo

ezyang committed:
Update on "[WIP] Split TensorIteratorConfig out of TensorIterator"


The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)

This diff doesn't compile, but I'm putting it up to get some early
feedback on the new API.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIteratorConfig config;
config.config1();
config.config2();
config.config3();
TensorIterator iter(config);
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@mruberry @ngimel I updated the PR to avoid copying TensorIteratorConfig into TensorIterator (it turns out, we can just discard most of the config operations after constructions), and avoid refcount bumps when we construct tensors and then operands again.

ezyang committed:
Update on "[WIP] Split TensorIteratorConfig out of TensorIterator"


The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)
In fact, most of the stuff in TensorIteratorConfig isn't used by
TensorIterator later, so it can be dropped entirely after construction.

This diff doesn't compile, but I'm putting it up to get some early
feedback on the new API.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIteratorConfig config;
config.config1();
config.config2();
config.config3();
TensorIterator iter(config);
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Split TensorIteratorConfig out of TensorIterator"


The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)
In fact, most of the stuff in TensorIteratorConfig isn't used by
TensorIterator later, so it can be dropped entirely after construction.

This diff doesn't compile, but I'm putting it up to get some early
feedback on the new API.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIteratorConfig config;
config.config1();
config.config2();
config.config3();
TensorIterator iter(config);
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22018845](https://our.internmc.facebook.com/intern/diff/D22018845)

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Split TensorIteratorConfig out of TensorIterator"


The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)
In fact, most of the stuff in TensorIteratorConfig isn't used by
TensorIterator later, so it can be dropped entirely after construction.

This diff doesn't compile, but I'm putting it up to get some early
feedback on the new API.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIteratorConfig config;
config.config1();
config.config2();
config.config3();
TensorIterator iter(config);
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22018845](https://our.internmc.facebook.com/intern/diff/D22018845)

[ghstack-poisoned]

VitalyFedyunin commented:
Please trigger internal Framework benchmarks after you are done with code changes.

ezyang commented:
Use sites are all updated now!

ngimel commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -65,18 +65,14 @@ struct DimCounter {
   OperandInfo() {}
-  explicit OperandInfo(const Tensor& t) : tensor(t) {
+  explicit OperandInfo(Tensor t) : tensor(std::move(t)) {
     if (t.defined()) {
```
use `tensor` here as `t` was moved from?

mruberry commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -38,7 +38,7 @@ void TensorIterator::reorder_dimensions() {
       }
       int64_t stride0 = operands_[arg].stride_bytes[dim0];
       int64_t stride1 = operands_[arg].stride_bytes[dim1];
-      if (is_reduction() && operands_[arg].is_output) {
```
Is it weird that is_reduction_ isn't queried from config?

ezyang replied:
Since the reduction parameter is copied in from config, it doesn't really matter

mruberry commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -670,66 +670,60 @@ void TensorIterator::select_all_keeping_dim(int start_dim, IntArrayRef indices)
 
 TensorIterator TensorIterator::binary_op(Tensor& out, const Tensor& a,
     const Tensor& b, bool check_mem_overlap) {
-  auto iter = TensorIterator();
```
Non-fluent?

mruberry commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -670,66 +670,60 @@ void TensorIterator::select_all_keeping_dim(int start_dim, IntArrayRef indices)
 
 TensorIterator TensorIterator::comparison_op(Tensor& out, const Tensor& a,
     const Tensor& b, bool check_mem_overlap) {
-  auto iter = TensorIterator();
```
Also non-fluent

mruberry commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -670,66 +670,60 @@ void TensorIterator::select_all_keeping_dim(int start_dim, IntArrayRef indices)
 
 TensorIterator TensorIterator::unary_op(Tensor& out, const Tensor& a,
     bool check_mem_overlap) {
-  auto iter = TensorIterator();
```
Make fluent?

mruberry commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -670,66 +670,60 @@ void TensorIterator::select_all_keeping_dim(int start_dim, IntArrayRef indices)
 }
 
 TensorIterator TensorIterator::nullary_op(Tensor& out) {
-  auto iter = TensorIterator();
```
Fluency

mruberry commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -670,66 +670,60 @@ void TensorIterator::select_all_keeping_dim(int start_dim, IntArrayRef indices)
 
 TensorIterator TensorIterator::reduce_op(Tensor& out, const Tensor& a) {
   TORCH_INTERNAL_ASSERT(out.defined());
-  auto iter = TensorIterator();
```
Fluent or not

mruberry commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -744,18 +738,25 @@ TensorIterator TensorIterator::reduce_op(Tensor& out1, Tensor& out2, const Tenso
       " and output2 has ", out2.sizes());
   TORCH_CHECK(out1.strides() == out2.strides(), "reduce_op(): expected both outputs to have same strides, but output1 has ", out1.strides(),
       " and output2 has ", out2.strides());
-  auto iter = TensorIterator();
```
Fluency

mruberry commented on aten/src/ATen/test/tensor_iterator_test.cpp:
```diff
@@ -166,47 +166,47 @@ TEST(TensorIteratorTest, SerialLoopSingleThread) {
 
 TEST(TensorIteratorTest, FailNonPromotingBinaryOp) {
   Tensor out;
-  auto iter = at::TensorIterator();
```
Non-fluent

mruberry reviewed (approved):
Added some style nits. @ngimel's issue is the only blocker I saw. (Well, that and the failing tests.)

ezyang commented:
@mruberry in some cases things cannot be made fluent, as I need to loop over something and do a variable number of calls to the fluent interface. Easiest to just directly allocate a config object in those locations.

ezyang committed:
Update on "Split TensorIteratorConfig out of TensorIterator"


The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)
In fact, most of the stuff in TensorIteratorConfig isn't used by
TensorIterator later, so it can be dropped entirely after construction.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIterator iter = TensorIteratorConfig()
  .config1()
  .config2()
  .config3()
  .build();
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22018845](https://our.internmc.facebook.com/intern/diff/D22018845)

[ghstack-poisoned]

ezyang committed:
Update on "Split TensorIteratorConfig out of TensorIterator"


The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)
In fact, most of the stuff in TensorIteratorConfig isn't used by
TensorIterator later, so it can be dropped entirely after construction.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIterator iter = TensorIteratorConfig()
  .config1()
  .config2()
  .config3()
  .build();
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22018845](https://our.internmc.facebook.com/intern/diff/D22018845)

[ghstack-poisoned]

ezyang commented:
It turns out I can't actually store references to input tensors, since there are multiple cases where people specify temporaries as inputs to the builder (and if build() doesn't happen in the same expression, they're invalid). So now I store full-blown owned references and then move them out of the config.

ngimel commented:
Hm, is it possible to change the contract so that references have to remain valid for the lifetime of TI, and change the callsites? Currently TI owns the references too, so it's not a regression, but would be cool to avoid it.

ezyang committed:
Update on "Split TensorIteratorConfig out of TensorIterator"


The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)
In fact, most of the stuff in TensorIteratorConfig isn't used by
TensorIterator later, so it can be dropped entirely after construction.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIterator iter = TensorIteratorConfig()
  .config1()
  .config2()
  .config3()
  .build();
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22018845](https://our.internmc.facebook.com/intern/diff/D22018845)

[ghstack-poisoned]

ezyang commented:
> Hm, is it possible to change the contract so that references have to remain valid for the lifetime of TI, and change the callsites?

Yes, this is possible. I'll have to think about how to write the API so that it's difficult for people to mess it up, however!

ezyang committed:
Update on "Split TensorIteratorConfig out of TensorIterator"


The basic concept is to make it more clear what the construction
side API is, as opposed to the "I want to actually do kernel stuff
with TensorIterator" API (which has been kept on TensorIterator.)
In fact, most of the stuff in TensorIteratorConfig isn't used by
TensorIterator later, so it can be dropped entirely after construction.

Before:

```
TensorIterator iter;
iter.config1();
iter.config2();
iter.config3();
iter.build();
// use iter
```

Now:

```
TensorIterator iter = TensorIteratorConfig()
  .config1()
  .config2()
  .config3()
  .build();
// use iter
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22018845](https://our.internmc.facebook.com/intern/diff/D22018845)

[ghstack-poisoned]
